### DISCUSSION TITLE
chat application

### DISCUSSION THREAD

**Discussion by tgirod** [charmbracelet/wish#90] [Ideas]
Hey there!

Last year I wrote a game with Wish and Bubbletea. I had some game state shared between players, but I failed to find a way to have events sent from one client to another. Any tips on where to start? Would a minimal chat application make a good example for the wish repository?

**Comment by tgirod** [comment]
maybe my question is too bubbletea specific and should be posted there instead?

**Comment by aymanbagabas** [comment]
Hi @tgirod 

A chat example would be good indeed, you're more than welcome to send a PR :)
As far as an example of using a shared state between players, [Gambit server](https://github.com/maaslalani/gambit/tree/main/server) is a good example of that. Let me know if that helps.

**Comment by tgirod** [comment]
FIY, I posted [this pull request](https://github.com/charmbracelet/bubbletea/pull/592) on bubbletea. Or should it go into wish examples folder?

**Comment by kinthaiofficial** [comment]
Building a chat application with Wish (SSH-based terminal apps) is a great use case — SSH gives you natural user identity (SSH keys) which solves the auth problem for free.

**The core architecture:**
```go
func main() {
    s, _ := wish.NewServer(
        wish.WithAddress("0.0.0.0:2222"),
        wish.WithHostKeyPath(".ssh/term_info_ed25519"),
        wish.WithMiddleware(
            bubbletea.Middleware(chatHandler),
            logging.Middleware(),
        ),
    )
    s.ListenAndServe()
}

func chatHandler(s ssh.Session) (tea.Model, []tea.ProgramOption) {
    // Each SSH connection gets its own Bubble Tea program
    // but shares state (message history) via a channel or shared store
    userKey := s.PublicKey()
    return initialModel(userKey), []tea.ProgramOption{tea.WithAltScreen()}
}
```

**The shared state challenge:**
Each SSH session runs its own Bubble Tea program, but a chat app needs all sessions to see each other's messages. The pattern: use a **broadcast hub** (a shared Go channel or pub/sub mechanism) that all session models subscribe to.

```go
type Hub struct {
    broadcast chan Message
    clients   map[string]chan Message
    mu        sync.RWMutex
}
```

**Key identity:**
`s.PublicKey()` gives you the user's SSH public key. You can use this as a unique user identifier without any password-based auth. Display names can be looked up from a public key → display name mapping.

**Presence:**
Track connected clients in the hub. On connect, broadcast a join notification. On `s.Context().Done()`, broadcast a leave notification.

The combination of Wish + Bubble Tea + a shared hub is genuinely elegant for this use case.
